File: backend/app/schemas/match.py
import datetime
import uuid
from typing import Any, Literal
from pydantic import BaseModel, ConfigDict, Field


MatchCategory = Literal["STRONG_MATCH", "GOOD_MATCH", "REVIEW", "LOW_MATCH"]


class MatchScores(BaseModel):
    skill: float | None = None
    role: float | None = None
    experience: float | None = None
    project: float | None = None
    education: float | None = None
    location: float | None = None
    semantic: float | None = None


class RelevantProject(BaseModel):
    project: str
    reason: str


class MatchExplanationResponse(BaseModel):
    strengths: list[str] = Field(default_factory=list)
    gaps: list[str] = Field(default_factory=list)
    explanation: str = ""
    relevant_projects: list[RelevantProject] = Field(default_factory=list)


class MatchResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    job_id: uuid.UUID
    candidate_profile_id: uuid.UUID
    overall_score: float
    category: MatchCategory
    match_version: str
    scores: MatchScores
    matching_skills: list[str]
    matching_preferred_skills: list[str]
    missing_required_skills: list[str]
    role_alignment: dict[str, Any]
    experience_alignment: dict[str, Any]
    education_alignment: dict[str, Any]
    project_alignment: dict[str, Any]
    location_alignment: dict[str, Any]
    strengths: list[str]
    gaps: list[str]
    explanation: str | None
    relevant_projects: list[RelevantProject]
    is_stale: bool
    created_at: datetime.datetime
    updated_at: datetime.datetime


class JobMatchListItem(BaseModel):
    id: uuid.UUID
    job_id: uuid.UUID
    job_title: str
    company: str
    location: str | None
    work_mode: str | None
    overall_score: float
    category: MatchCategory
    is_stale: bool
    matching_skills: list[str]
    missing_required_skills: list[str]
    role_alignment: dict[str, Any]
    created_at: datetime.datetime


class PaginatedMatches(BaseModel):
    items: list[JobMatchListItem]
    page: int
    page_size: int
    total: int
    total_pages: int

File: frontend/types/match.ts
export type MatchCategory = "STRONG_MATCH" | "GOOD_MATCH" | "REVIEW" | "LOW_MATCH";

export interface MatchScores {
  skill: number | null;
  role: number | null;
  experience: number | null;
  project: number | null;
  education: number | null;
  location: number | null;
  semantic: number | null;
}

export interface AlignmentBlock {
  status?: string | null;
  reason?: string | null;
  weight?: number | null;
  required?: string | null;
  candidate_years?: number | null;
  relevant_count?: number | null;
  job_title?: string | null;
  target_role?: string | null;
}

export interface RelevantProject {
  project: string;
  reason: string;
  score: number;
}

export interface MatchResponse {
  id: string;
  job_id: string;
  candidate_profile_id: string;
  overall_score: number;
  category: MatchCategory;
  match_version: string;
  scores: MatchScores;
  matching_skills: string[];
  matching_preferred_skills: string[];
  related_required_matches: string[];
  missing_required_skills: string[];
  role_alignment: AlignmentBlock;
  experience_alignment: AlignmentBlock;
  education_alignment: AlignmentBlock;
  project_alignment: AlignmentBlock;
  location_alignment: AlignmentBlock;
  strengths: string[];
  gaps: string[];
  explanation: string | null;
  relevant_projects: RelevantProject[];
  is_stale: boolean;
  created_at: string;
  updated_at: string;
}

export interface JobMatchListItem {
  id: string;
  job_id: string;
  job_title: string;
  company: string;
  location: string | null;
  work_mode: string | null;
  overall_score: number;
  category: MatchCategory;
  is_stale: boolean;
  matching_skills: string[];
  missing_required_skills: string[];
  role_alignment: AlignmentBlock;
  created_at: string;
}

export interface PaginatedMatches {
  items: JobMatchListItem[];
  page: number;
  page_size: number;
  total: number;
  total_pages: number;
}

export interface MatchFilters {
  page?: number;
  page_size?: number;
  sort?: string;
  category?: string;
  score_min?: number;
  score_max?: number;
  location?: string;
  role?: string;
  company?: string;
  stale?: boolean;
}

export const MATCH_CATEGORIES: MatchCategory[] = ["STRONG_MATCH", "GOOD_MATCH", "REVIEW", "LOW_MATCH"];
export const MATCH_SORTS = ["match", "newest", "oldest", "role", "company", "stale"] as const;

export function categoryRank(category: MatchCategory | string | undefined | null): number {
  const index = MATCH_CATEGORIES.indexOf(category as MatchCategory);
  return index === -1 ? MATCH_CATEGORIES.length : index;
}